File: Programas/NoInformados/Rompecabezas/BFS/Scripts/BFS.js

export function obtenerNumeroNodos(nodoInicial) {
    if (nodoInicial.hijos == 0)
        return 1
    let hijos = 0
    for (let i = 0; i < nodoInicial.hijos.length; i++) {
        hijos = hijos + obtenerNumeroNodos(nodoInicial.hijos[i])
    }
    return hijos + 1
}

export function obtenerNumeroRamas(nodoInicial) {
    if (nodoInicial.hijos == 0)
        return 1
    let hijos = 0
    for (let i = 0; i < nodoInicial.hijos.length; i++) {
        hijos = hijos + obtenerNumeroRamas(nodoInicial.hijos[i])
    }
    return hijos
}

// Algoritmo Busqueda en Profundidad
export class BFS {
    static buscar(nodoInicial, nodoFinal) {
        let agenda = []
        let configuracionesExploradas = new Set()

        if (nodoFinal.compararCeldas(nodoInicial)){
            nodoInicial.objetivo = true
            return nodoInicial
        }

        agenda.push(nodoInicial)

        while (agenda.length > 0) {

            let nodo = agenda.shift()

            configuracionesExploradas.add(nodo.celdas)

            for (let hijo of nodo.expandir()) {

                if (nodoFinal.compararCeldas(hijo)){
                    hijo.objetivo = true
                    return hijo
                }
                else if (!configuracionesExploradas.has(hijo.celdas))
                    agenda.push(hijo)
            }
        }

        return null
    }
}

File: Programas/NoInformados/Rompecabezas/2. Busqueda en Amplitud/C++ BFS/BFS.cpp
#include "./Header/BFS.h"

std::vector<Rompecabezas*> BFS::trayectoria( Rompecabezas * nodoFinal ){
    std::vector<Rompecabezas*> secuencia;
    secuencia.push_back( nodoFinal );
    while( nodoFinal->nodopadre != nullptr ){
        nodoFinal = nodoFinal->nodopadre;
        secuencia.push_back( nodoFinal );
    }
    return secuencia;
}

std::vector<Rompecabezas*> BFS::Bucar( Rompecabezas * nodoInicial , Rompecabezas * nodoFinal ){
    std::vector<Rompecabezas*> agenda;
    std::set<unsigned long long> configuracionesExploradas;

    if( nodoFinal->compararCeldas(nodoInicial) )
        return BFS::trayectoria(nodoInicial);
        
    agenda.push_back( nodoInicial );

    while( agenda.size() > 0 ){
        Rompecabezas * nodo = agenda[0];
        agenda.erase( agenda.begin() );

        configuracionesExploradas.insert( nodo->tablero );
        for( Rompecabezas * hijo : nodo->expandirNodos() ){
            if( nodoFinal->compararCeldas(hijo) )
                return BFS::trayectoria( hijo );
            else if( !configuracionesExploradas.count( hijo->tablero ) )
                agenda.push_back( hijo );
        }
    }

    return BFS::trayectoria( new Rompecabezas( nullptr , ' ' , 0 ) );
}